// @taquito/taquito/dist/lib/contract/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidViewSimulationContext = exports.validateAndExtractFailwith = exports.ViewSimulationError = exports.InvalidViewParameterError = exports.InvalidInitParameter = exports.InvalidCodeParameter = exports.InvalidDelegationSource = exports.UndefinedLambdaContractError = exports.InvalidParameterError = void 0;
var InvalidParameterError = /** @class */ (function () {
    function InvalidParameterError(smartContractMethodName, sigs, args) {
        this.smartContractMethodName = smartContractMethodName;
        this.sigs = sigs;
        this.args = args;
        this.name = 'Invalid parameters error';
        this.message = smartContractMethodName + " Received " + args.length + " arguments while expecting one of the following signatures (" + JSON.stringify(sigs) + ")";
    }
    return InvalidParameterError;
}());
exports.InvalidParameterError = InvalidParameterError;
var UndefinedLambdaContractError = /** @class */ (function () {
    function UndefinedLambdaContractError() {
        this.name = 'Undefined LambdaContract error';
        this.message = "This might happen if you are using a sandbox. Please provide the address of a lambda contract as a parameter of the read method.";
    }
    return UndefinedLambdaContractError;
}());
exports.UndefinedLambdaContractError = UndefinedLambdaContractError;
var InvalidDelegationSource = /** @class */ (function () {
    function InvalidDelegationSource(source) {
        this.source = source;
        this.name = 'Invalid delegation source error';
        this.message = "Since Babylon delegation source can no longer be a contract address " + source + ". Please use the smart contract abstraction to set your delegate.";
    }
    return InvalidDelegationSource;
}());
exports.InvalidDelegationSource = InvalidDelegationSource;
var InvalidCodeParameter = /** @class */ (function () {
    function InvalidCodeParameter(message, data) {
        this.message = message;
        this.data = data;
        this.name = 'InvalidCodeParameter';
    }
    return InvalidCodeParameter;
}());
exports.InvalidCodeParameter = InvalidCodeParameter;
var InvalidInitParameter = /** @class */ (function () {
    function InvalidInitParameter(message, data) {
        this.message = message;
        this.data = data;
        this.name = 'InvalidInitParameter';
    }
    return InvalidInitParameter;
}());
exports.InvalidInitParameter = InvalidInitParameter;
var InvalidViewParameterError = /** @class */ (function () {
    function InvalidViewParameterError(smartContractViewName, sigs, args, originalError) {
        this.smartContractViewName = smartContractViewName;
        this.sigs = sigs;
        this.args = args;
        this.originalError = originalError;
        this.name = 'Invalid view parameters error';
        this.message = "Unable to encode the parameter of the view: " + smartContractViewName + ". Received " + args + " as parameter while expecting one of the following signatures (" + JSON.stringify(sigs) + ")";
        this.cause = originalError;
    }
    return InvalidViewParameterError;
}());
exports.InvalidViewParameterError = InvalidViewParameterError;
var ViewSimulationError = /** @class */ (function () {
    function ViewSimulationError(message, originalError) {
        this.message = message;
        this.originalError = originalError;
        this.name = 'ViewSimulationError';
    }
    return ViewSimulationError;
}());
exports.ViewSimulationError = ViewSimulationError;
var validateAndExtractFailwith = function (error) {
    if (isJsonString(error.body)) {
        var parsedError = JSON.parse(error.body);
        if (Array.isArray(parsedError) && 'with' in parsedError[parsedError.length - 1]) {
            return parsedError[parsedError.length - 1].with;
        }
    }
};
exports.validateAndExtractFailwith = validateAndExtractFailwith;
var isJsonString = function (str) {
    try {
        JSON.parse(str);
    }
    catch (e) {
        return false;
    }
    return true;
};
var InvalidViewSimulationContext = /** @class */ (function () {
    function InvalidViewSimulationContext(info) {
        this.info = info;
        this.name = 'InvalidViewSimulationContext';
        this.message = info + " Please configure the context of the view execution in the executeView method.";
    }
    return InvalidViewSimulationContext;
}());
exports.InvalidViewSimulationContext = InvalidViewSimulationContext;
