// @taquito/michelson-encoder/dist/lib/tokens/chest-key.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChestKeyToken = exports.ChestKeyValidationError = void 0;
var token_1 = require("./token");
var ChestKeyValidationError = /** @class */ (function (_super) {
    __extends(ChestKeyValidationError, _super);
    function ChestKeyValidationError(value, token, message) {
        var _this = _super.call(this, value, token, message) || this;
        _this.value = value;
        _this.token = token;
        _this.name = 'ChestKeyValidationError';
        return _this;
    }
    return ChestKeyValidationError;
}(token_1.TokenValidationError));
exports.ChestKeyValidationError = ChestKeyValidationError;
var ChestKeyToken = /** @class */ (function (_super) {
    __extends(ChestKeyToken, _super);
    function ChestKeyToken(val, idx, fac) {
        var _this = _super.call(this, val, idx, fac) || this;
        _this.val = val;
        _this.idx = idx;
        _this.fac = fac;
        return _this;
    }
    ChestKeyToken.prototype.isValid = function (val) {
        if (/^[0-9a-fA-F]*$/.test(val) && val.length % 2 === 0) {
            return null;
        }
        else {
            return new ChestKeyValidationError(val, this, "Invalid bytes: " + val);
        }
    };
    ChestKeyToken.prototype.convertUint8ArrayToHexString = function (val) {
        return val.constructor === Uint8Array ? Buffer.from(val).toString('hex') : val;
    };
    ChestKeyToken.prototype.Encode = function (args) {
        var val = args.pop();
        val = this.convertUint8ArrayToHexString(val);
        var err = this.isValid(val);
        if (err) {
            throw err;
        }
        return { bytes: val };
    };
    ChestKeyToken.prototype.EncodeObject = function (val) {
        val = this.convertUint8ArrayToHexString(val);
        var err = this.isValid(val);
        if (err) {
            throw err;
        }
        return { bytes: val };
    };
    ChestKeyToken.prototype.Execute = function (val) {
        return val.bytes;
    };
    ChestKeyToken.prototype.ExtractSchema = function () {
        return ChestKeyToken.prim;
    };
    ChestKeyToken.prototype.findAndReturnTokens = function (tokenToFind, tokens) {
        if (ChestKeyToken.prim === tokenToFind) {
            tokens.push(this);
        }
        return tokens;
    };
    ;
    ChestKeyToken.prim = 'chest_key';
    return ChestKeyToken;
}(token_1.Token));
exports.ChestKeyToken = ChestKeyToken;
